### DISCUSSION TITLE
Ver 1.5 and greater MR mode bit clarification

### DISCUSSION THREAD

**Discussion by swelch** [ofiwg/libfabric#7878] [General]
I'm looking for a clarification with regard to mr_mode set in hints for >= 1.5 version. The reading of the MR mode bits description for fi_domain(3) suggest:
_Applications that require the use of a specific registration mode should set the domain attribute mr_mode to the necessary value when calling fi_getinfo. The value FI_MR_UNSPEC may be used to indicate support for any registration mode._

However, in the routine ofi_check_mr_mode(), prov/util/src/util_attr.c line 546, an exact match between hints->mr_mode and prov->mr_mode is required to match. Which is the correct intent, and if the documentation is correct shouldn't this check be something more like this:

```
545                         prov_mode = ofi_cap_mr_mode(user_info->caps, prov_mode);
546                         if (user_mode != FI_MR_UNSPEC &&
547                             (user_mode & prov_mode) != prov_mode)
548                                         goto out;
```

**Comment by j-xiong** [comment]
Current code at L546 `if ((user_mode & prov_mode) != prov_mode)` doesn't require exact match - it requires user_mode be explicit and meet  or exceed provider requirement. 

I am not sure  why FI_MR_UNSPEC is not handled here. It should be allowed.

**Reply by swelch** [reply]
> Current code at L546 `if ((user_mode & prov_mode) != prov_mode)` doesn't require exact match - it requires user_mode be explicit and meet or exceed provider requirement.
> 
Agreed, thanks.

> I am not sure why FI_MR_UNSPEC is not handled here. It should be allowed.

I will open a PR.
